File: 0202 Happy Number/0202.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def next_num(num):
            return sum([int(c)**2 for c in str(num)])

        fast, slow = next_num(n), n

        while fast != 1 and fast != slow:
            fast = next_num(next_num(fast))
            slow = next_num(slow)
        
        return fast == 1

File: 0202 Happy Number/0202.cpp
class Solution {
public:
    int next(int num){
        int nextNum = 0;

        while (num > 0){
            int cur = num % 10;
            nextNum += cur * cur;
            num /= 10;
        }

        return nextNum; 
    }

    bool isHappy(int n) {
        int fast = next(n);
        int slow = n;

        while (fast != 1 && fast != slow){
            slow = next(slow);
            fast = next(next(fast));
        }

        return fast == 1;
    }
};
